File: app/src/main/java/io/github/myolwin00/sunshine/data/viewmodel/WeatherViewModelFactory.java
package io.github.myolwin00.sunshine.data.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProvider;

import io.github.myolwin00.sunshine.data.repository.WeatherRepository;

/**
 * Created by emrys on 2/19/18.
 */

public class WeatherViewModelFactory implements ViewModelProvider.Factory {

    private WeatherRepository mRepository;

    public WeatherViewModelFactory(WeatherRepository repository) {
        mRepository = repository;
    }

    @NonNull
    @Override
    public ForecastVM create(@NonNull Class modelClass) {
        return new ForecastVM(mRepository);
    }
}

File: app/src/main/java/io/github/myolwin00/sunshine/data/viewmodel/WeatherViewModelFactory.kt
package io.github.myolwin00.sunshine.data.viewmodel

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import io.github.myolwin00.sunshine.data.repository.WeatherRepository
import javax.inject.Inject
import javax.inject.Singleton

import javax.inject.Provider


/**
 * Created by emrys on 1/19/18.
 */
@Suppress("UNCHECKED_CAST")
@Singleton
class ViewModelFactory @Inject
constructor(
        private val creators: Map<Class<out ViewModel>, @JvmSuppressWildcards Provider<ViewModel>>
): ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        var creator: Provider<out ViewModel>? = creators[modelClass]
        if (creator == null) {
            for ((key, value) in creators) {
                if (modelClass.isAssignableFrom(key)) {
                    creator = value
                    break
                }
            }
        }
        if (creator == null) {
            throw IllegalArgumentException("unknown model class " + modelClass)
        }
        try {
            return creator.get() as T
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

    }
}